File: Stack/735-asteroid-collision/asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stk= new Stack<>();

        for(int num: asteroids) {
            
           boolean alive = true;
           while(alive && num < 0 && !stk.isEmpty() && stk.peek() > 0) {
                int top = stk.peek();

                if(Math.abs(num) > top) {
                    stk.pop();
                } else if(Math.abs(num) == top) {
                    stk.pop();
                    alive = false;
                } else {
                    alive = false;
                }
           }
           if (alive) stk.push(num);
        }

        return stk.stream().mapToInt(Integer::intValue).toArray();


        
    }
}

File: Array/735-asteroid-collision/asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stk = []

        for ast in asteroids:
            
            while stk and ast < 0 and ast!=float('-inf') and stk[-1] > 0:
                top = stk.pop()
                if abs(top) > abs(ast):
                    ast = top
                elif abs(top) == abs(ast):
                    ast = float('-inf')
            
            if ast!=float('-inf'):
                stk.append(ast)
        return stk